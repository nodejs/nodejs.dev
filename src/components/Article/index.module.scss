.article {
  max-width: 768px;
  overflow-wrap: break-word;
  padding: var(--space-80) 0 var(--space-24) var(--space-128);
  width: calc(100% - 36rem);

  .headline {
    margin: 0 0 var(--space-40) 0;
    padding: 0;

    @media (max-width: 800px) {
      font-size: var(--font-size-display3);
      line-height: 1.2;
      margin: 6rem 0 2rem 0;
    }
  }

  @media (max-width: 800px) {
    box-sizing: border-box;
    max-width: 100vw;
    padding: 0 3rem;
    width: 100%;
  }

  img {
    max-width: 100%;
  }

  hr {
    border: none;
    border-bottom: 0.4rem solid var(--black2);
    margin-bottom: var(--space-40);
  }

  blockquote {
    border-left: 3px solid var(--black4);
    color: var(--color-text-primary);
    font-style: italic;
    font-weight: 200;
    letter-spacing: 1px;
    margin: 0 0 2.8rem 0;
    padding: 1.8rem 0 1.8rem 3.2rem;

    p {
      margin: 0;
      padding: 0;
    }
  }

  section {
    margin-bottom: var(--space-48);
  }

  h4 {
    margin-bottom: var(--space-16);
  }

  li {
    + h4 {
      margin-top: var(--space-32);
    }

    margin-bottom: var(--space-08);
  }

  h2 {
    scroll-margin-top: var(--space-128);
  }

  h3 {
    margin-bottom: var(--space-24);
    scroll-margin-top: var(--space-128);
  }

  details,
  iframe {
    margin-bottom: var(--space-24);
  }

  p + {
    h2 {
      margin-top: var(--space-40);
    }

    h3 {
      margin-top: var(--space-32);
    }
  }
}
