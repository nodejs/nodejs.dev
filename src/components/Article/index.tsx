import React from 'react';
import { throttle } from 'throttle-debounce';
import { PaginationInfo } from '../../types';
import AuthorsList from '../../containers/AuthorList';
import EditLink from '../EditLink';
import Pagination from '../Pagination';
import TOC from '../Toc';

interface Props {
  title: string;
  html: string;
  tableOfContents: string;
  authors: string[];
  relativePath?: string;
  editPath?: string;
  next?: PaginationInfo;
  previous?: PaginationInfo;
}

const NAV_HEIGHT = 72;

const Article = ({
  title,
  html,
  tableOfContents,
  previous,
  next,
  relativePath,
  editPath,
  authors,
}: Props): JSX.Element => {
  const element = React.useRef<HTMLElement | null>(null);

  const handleRef = (ref?: HTMLElement | null): void => {
    if (ref) {
      element.current = ref;
    }
  };

  React.useEffect((): (() => void) => {
    let observer: IntersectionObserver;

    if (window.history.state && window.history.state.articleScrollTo) {
      window.scrollTo({
        top: window.history.state.articleScrollTo,
      });
    }

    if (element.current) {
      const handleObserverThrottled = throttle(
        300,
        (entries: IntersectionObserverEntry[]) => {
          entries.forEach((entry): void => {
            // element is already hidden by the nav
            if (entry.boundingClientRect.y < NAV_HEIGHT) {
              if (!entry.target.previousElementSibling) {
                window.history.replaceState(
                  {
                    articleScrollTo: null,
                  },
                  '',
                  null
                );
                return;
              }

              window.history.replaceState(
                {
                  articleScrollTo: document.documentElement.scrollTop,
                },
                '',
                null
              );
            }
          });
        }
      );

      observer = new IntersectionObserver(
        (entries): void => {
          handleObserverThrottled(entries);
        },
        {
          threshold: [0.25, 0.5, 0.75],
          rootMargin: `-${NAV_HEIGHT}px 0px 0px 0px`,
        }
      );

      Array.from(element.current.children).forEach((children): void => {
        observer.observe(children);
      });
    }

    return (): void => {
      if (observer && element.current) {
        Array.from(element.current.children).forEach((children): void => {
          observer.unobserve(children);
        });
      }
    };
  }, []);

  return (
    <article className="article-reader">
      <h1 className="article-reader__headline">{title}</h1>
      <TOC heading="TABLE OF CONTENTS" tableOfContents={tableOfContents} />
      {/* eslint-disable-next-line react/no-danger */}
      <div ref={handleRef} dangerouslySetInnerHTML={{ __html: html }} />
      <AuthorsList authors={authors} />
      <EditLink relativePath={relativePath} editPath={editPath} />
      <Pagination previous={previous} next={next} />
    </article>
  );
};

export default Article;
