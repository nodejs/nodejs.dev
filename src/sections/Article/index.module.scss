.article {
  max-width: 768px;
  overflow-wrap: break-word;
  padding: var(--space-48) var(--space-32) var(--space-48) var(--space-32);
  width: calc(100vw - 36rem);

    background: url(/static/images/header.svg);
    background-repeat: no-repeat;
    background-size:contain;
    
  .headline {
    margin: 0 0 var(--space-40) 0;
    padding: 0;

    @media (max-width: 600px) {
      font-size: var(--font-size-display3);
      line-height: 1.2;
      margin: var(--space-24) 0 var(--space-24) 0;
    }
  }

  @media (min-width: 1400px) {
    max-width: 980px;
    padding: var(--space-80) var(--space-32) var(--space-48) var(--space-128);
  }

  @media (max-width: 900px) {
    padding: var(--space-80) var(--space-32) var(--space-48) var(--space-32);
    width: auto;
  }

  @media (max-width: 600px) {
    box-sizing: border-box;
    max-width: 100vw;
    padding: var(--space-24) 3rem;
    width: 100%;
  }

  img {
    max-width: 100%;
  }

  hr {
    border: none;
    border-bottom: 0.4rem solid var(--black2);
    margin-bottom: var(--space-40);
  }

  blockquote {
    border-left: 3px solid var(--black4);
    color: var(--color-text-primary);
    font-style: italic;
    font-weight: var(--font-weight-light);
    letter-spacing: 1px;
    margin: 0 0 2.8rem 0;
    padding: 1.8rem 0 1.8rem 3.2rem;

    p {
      margin: 0;
      padding: 0;
    }
  }

  section {
    margin-bottom: var(--space-48);
  }

  h4 {
    margin-bottom: var(--space-16);
  }

  li {
    + h4 {
      margin-top: var(--space-32);
    }

    margin-bottom: var(--space-08);
  }

  h2 {
    scroll-margin-top: var(--space-128);
  }

  h3 {
    margin-bottom: var(--space-24);
    scroll-margin-top: var(--space-128);
  }

  details,
  iframe {
    margin-bottom: var(--space-24);
  }

  p + {
    h2 {
      margin-top: var(--space-40);
    }

    h3 {
      margin-top: var(--space-32);
    }
  }
}