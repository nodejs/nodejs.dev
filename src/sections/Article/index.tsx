import React from 'react';
import { MDXRenderer } from 'gatsby-plugin-mdx';
import { MDXProvider } from '@mdx-js/react';
import { BlogComponents, ArticleComponents } from '../../components';
import { mdxComponents } from '../../mdxComponents';
import {
  PaginationInfo,
  BlogPostAuthor,
  TableOfContentsItem,
} from '../../types';
import styles from './index.module.scss';

interface Props {
  title: string;
  description?: string;
  body: string;
  authors: string[] | BlogPostAuthor[];
  tableOfContents: TableOfContentsItem[];
  relativePath?: string;
  absolutePath?: string;
  editPath?: string;
  next?: PaginationInfo;
  previous?: PaginationInfo;
  blog?: boolean;
  date?: string;
  extraLinks?: JSX.Element[];
  children?: React.ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  extraComponents?: Record<string, (...any: any[]) => JSX.Element | null>;
  childrenPosition?: 'before' | 'after';
}

const renderBlogAuthors = (
  date: string | undefined,
  authors: BlogPostAuthor[] | undefined
) => <BlogComponents.BlogAuthorsList date={date} authors={authors || []} />;

const renderArticleAuthors = (authors: string[]) => (
  <ArticleComponents.AuthorList authors={authors || []} />
);

const renderTOC = (tableOfContents: TableOfContentsItem[] | undefined) => (
  <ArticleComponents.TableOfContents tableOfContents={tableOfContents || []} />
);

const Article = ({
  title,
  description,
  body,
  tableOfContents,
  previous,
  next,
  relativePath,
  absolutePath,
  editPath,
  authors,
  blog,
  date,
  children,
  extraLinks = [],
  extraComponents = {},
  childrenPosition = 'after',
}: Props): JSX.Element => (
  <article className={styles.article}>
    {childrenPosition === 'before' && children && <div>{children}</div>}
    <header>
      <h1 className={styles.headline}>{title}</h1>
      <p>{description}</p>
    </header>
    {blog
      ? renderBlogAuthors(date, authors as BlogPostAuthor[])
      : renderArticleAuthors(authors as string[])}
    {renderTOC(tableOfContents)}
    <div>
      <MDXProvider components={{ ...mdxComponents, ...extraComponents }}>
        <MDXRenderer>{body}</MDXRenderer>
      </MDXProvider>
    </div>
    {childrenPosition === 'after' && children && <div>{children}</div>}
    {!blog && (
      <>
        <ArticleComponents.EditLink
          absolutePath={absolutePath}
          relativePath={relativePath}
          editPath={editPath}
        />
        {extraLinks}
        <ArticleComponents.Pagination previous={previous} next={next} />
      </>
    )}
  </article>
);

export default Article;
