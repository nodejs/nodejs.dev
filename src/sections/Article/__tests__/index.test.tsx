import React from 'react';
import { render } from '@testing-library/react';
import Article from '..';
import {
  createLearnPageData,
  createLearnPageContext,
} from '../../../__fixtures__/page';

const getArticleProps = () => {
  const learnPageData = createLearnPageData();
  const learnPageContext = createLearnPageContext();

  const {
    data: { articleCurrentLanguage },
  } = learnPageData;

  const {
    frontmatter: { title },
    body,
    tableOfContents,
    fields: { authors },
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  } = articleCurrentLanguage!;

  const { relativePath, next, previous } = learnPageContext;

  return {
    title,
    body,
    next,
    previous,
    authors,
    relativePath,
    tableOfContents,
  };
};

describe('Article component', () => {
  const intersectionObserverOriginal = window.IntersectionObserver;
  const scrollToOriginal = window.scrollTo;
  const replaceStateOriginal = window.history.replaceState;

  afterEach(() => {
    window.IntersectionObserver = intersectionObserverOriginal;
    window.scrollTo = scrollToOriginal;
    window.history.replaceState = replaceStateOriginal;

    // clean-up history state
    window.history.replaceState({}, '', null);
  });

  it('renders correctly', () => {
    const {
      title,
      tableOfContents,
      body,
      next,
      previous,
      authors,
      relativePath,
    } = getArticleProps();

    const { container } = render(
      <Article
        title={title}
        tableOfContents={tableOfContents?.items || []}
        body={body}
        next={next}
        previous={previous}
        authors={authors}
        relativePath={relativePath}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders correctly in case body ref is null', () => {
    const {
      title,
      tableOfContents,
      body,
      next,
      previous,
      authors,
      relativePath,
    } = getArticleProps();

    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    jest.spyOn(React, 'useRef').mockReturnValueOnce(null);

    const { container } = render(
      <Article
        title={title}
        tableOfContents={tableOfContents?.items || []}
        body={body}
        next={next}
        previous={previous}
        authors={authors}
        relativePath={relativePath}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders correctly in blog mode', () => {
    const { title, tableOfContents, body, next, previous, relativePath } =
      getArticleProps();

    const authors = [
      {
        id: '1',
        name: 'test-user',
        website: '',
      },
    ];

    const { container } = render(
      <Article
        title={title}
        tableOfContents={tableOfContents?.items || []}
        body={body}
        next={next}
        previous={previous}
        authors={authors}
        relativePath={relativePath}
        blog
        date="date string from test"
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('should skip non-scrolled title from saving in history', () => {
    const {
      title,
      tableOfContents,
      body,
      next,
      previous,
      authors,
      relativePath,
    } = getArticleProps();

    window.history.replaceState = jest.fn();

    // mock IntersectionObserver
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    window.IntersectionObserver = jest.fn(callback => ({
      unobserve: jest.fn(),
      observe: jest.fn(() => {
        callback(
          [
            {
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              boundingClientRect: {
                y: 100, // title still in view, not scrolled
              },
              // eslint-disable-next-line @typescript-eslint/ban-ts-comment
              // @ts-ignore
              target: {
                previousElementSibling: null,
              },
            },
          ],
          jest.fn() as unknown as IntersectionObserver
        );
      }),
    }));

    render(
      <Article
        title={title}
        tableOfContents={tableOfContents?.items || []}
        body={body}
        next={next}
        previous={previous}
        authors={authors}
        relativePath={relativePath}
      />
    );

    expect(window.history.replaceState).toHaveBeenCalledTimes(0);
  });

  it('should accept and render child components', () => {
    const {
      title,
      tableOfContents,
      body,
      next,
      previous,
      authors,
      relativePath,
    } = getArticleProps();

    const { container } = render(
      <Article
        title={title}
        tableOfContents={tableOfContents?.items || []}
        body={body}
        next={next}
        previous={previous}
        authors={authors}
        relativePath={relativePath}
      >
        <div>test child JSX</div>
      </Article>
    );

    expect(container).toMatchSnapshot();
  });
});
