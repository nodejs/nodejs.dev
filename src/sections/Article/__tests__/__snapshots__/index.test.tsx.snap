// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article component renders correctly 1`] = `
<div>
  <article
    class="article"
  >
    <div
      class="header"
    >
      <h1
        class="headline"
      >
        test-title
      </h1>
      <p />
    </div>
    <div
      class="authorList"
    >
      Article Authors
      <ul>
        <li>
          <a
            aria-label="test-user1 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user1"
            rel="noopener noreferrer"
            style="margin-left: 0px;"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user1.png?size=60"
            />
          </a>
        </li>
        <li>
          <a
            aria-label="test-user2 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user2"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user2.png?size=60"
            />
          </a>
        </li>
      </ul>
    </div>
    <details
      class="tableOfContents"
    >
      <summary>
        <strong>
          Table of Contents
        </strong>
      </summary>
      <ul>
        <li>
          <a
            href="#mock-title"
          >
            mock-title
          </a>
        </li>
        <li>
          <a
            href="#mock-title-2"
          >
            mock-title-2
          </a>
        </li>
      </ul>
    </details>
    <div>
      <p>
        Node.js can be installed in different ways. This post highlights the most common and convenient ones.
      </p>
      <p>
        Official packages for all the major platforms are available at 
        <a>
          https://nodejs.dev/download/
        </a>
        .
      </p>
      <p>
        One very convenient way to install Node.js is through a package manager. In this case, every operating system has its own.
      </p>
      <p>
        Other package managers for MacOS, Linux, and Windows are listed in 
        <a>
          https://nodejs.dev/download/package-manager/
        </a>
      </p>
      <p>
        <code
          class="code"
        >
          nvm
        </code>
         is a popular way to run Node.js. It allows you to easily switch the Node.js version, and install new versions to try and easily rollback if something breaks, for example.
      </p>
      <p>
        It is also very useful to test your code with old Node.js versions.
      </p>
      <p>
        See 
        <a>
          https://github.com/nvm-sh/nvm
        </a>
         for more information about this option.
      </p>
      <p>
        In any case, when Node.js is installed you'll have access to the 
        <code
          class="code"
        >
          node
        </code>
         executable program in the command line.
      </p>
    </div>
    <div
      class="edit"
    >
      <a
        href="https://github.com/nodejs/nodejs.dev/edit/main/content/learn/test-path"
      >
        <span>
          Edit this page on GitHub
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pencil "
          data-icon="pencil"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
    <ul
      class="pagination"
    >
      <li>
        <a
          href="test-slug"
          rel="prev"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-left "
            data-icon="angles-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160zm352-160l-160 160c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L301.3 256 438.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0z"
              fill="currentColor"
            />
          </svg>
          Prev
        </a>
      </li>
      <li>
        <a
          href="test-slug"
          rel="next"
        >
          Next
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-right "
            data-icon="angles-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-left: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M470.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 256 265.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160zm-352 160l160-160c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L210.7 256 73.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0z"
              fill="currentColor"
            />
          </svg>
        </a>
      </li>
    </ul>
  </article>
</div>
`;

exports[`Article component renders correctly in blog mode 1`] = `
<div>
  <article
    class="article"
  >
    <div
      class="header"
    >
      <h1
        class="headline"
      >
        test-title
      </h1>
      <p />
    </div>
    <h5
      class="list"
    >
      date string from test
       by
       
      <span>
        test-user
      </span>
    </h5>
    <details
      class="tableOfContents"
    >
      <summary>
        <strong>
          Table of Contents
        </strong>
      </summary>
      <ul>
        <li>
          <a
            href="#mock-title"
          >
            mock-title
          </a>
        </li>
        <li>
          <a
            href="#mock-title-2"
          >
            mock-title-2
          </a>
        </li>
      </ul>
    </details>
    <div>
      <p>
        Node.js can be installed in different ways. This post highlights the most common and convenient ones.
      </p>
      <p>
        Official packages for all the major platforms are available at 
        <a>
          https://nodejs.dev/download/
        </a>
        .
      </p>
      <p>
        One very convenient way to install Node.js is through a package manager. In this case, every operating system has its own.
      </p>
      <p>
        Other package managers for MacOS, Linux, and Windows are listed in 
        <a>
          https://nodejs.dev/download/package-manager/
        </a>
      </p>
      <p>
        <code
          class="code"
        >
          nvm
        </code>
         is a popular way to run Node.js. It allows you to easily switch the Node.js version, and install new versions to try and easily rollback if something breaks, for example.
      </p>
      <p>
        It is also very useful to test your code with old Node.js versions.
      </p>
      <p>
        See 
        <a>
          https://github.com/nvm-sh/nvm
        </a>
         for more information about this option.
      </p>
      <p>
        In any case, when Node.js is installed you'll have access to the 
        <code
          class="code"
        >
          node
        </code>
         executable program in the command line.
      </p>
    </div>
  </article>
</div>
`;

exports[`Article component renders correctly in case body ref is null 1`] = `
<div>
  <article
    class="article"
  >
    <div
      class="header"
    >
      <h1
        class="headline"
      >
        test-title
      </h1>
      <p />
    </div>
    <div
      class="authorList"
    >
      Article Authors
      <ul>
        <li>
          <a
            aria-label="test-user1 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user1"
            rel="noopener noreferrer"
            style="margin-left: 0px;"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user1.png?size=60"
            />
          </a>
        </li>
        <li>
          <a
            aria-label="test-user2 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user2"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user2.png?size=60"
            />
          </a>
        </li>
      </ul>
    </div>
    <details
      class="tableOfContents"
    >
      <summary>
        <strong>
          Table of Contents
        </strong>
      </summary>
      <ul>
        <li>
          <a
            href="#mock-title"
          >
            mock-title
          </a>
        </li>
        <li>
          <a
            href="#mock-title-2"
          >
            mock-title-2
          </a>
        </li>
      </ul>
    </details>
    <div>
      <p>
        Node.js can be installed in different ways. This post highlights the most common and convenient ones.
      </p>
      <p>
        Official packages for all the major platforms are available at 
        <a>
          https://nodejs.dev/download/
        </a>
        .
      </p>
      <p>
        One very convenient way to install Node.js is through a package manager. In this case, every operating system has its own.
      </p>
      <p>
        Other package managers for MacOS, Linux, and Windows are listed in 
        <a>
          https://nodejs.dev/download/package-manager/
        </a>
      </p>
      <p>
        <code
          class="code"
        >
          nvm
        </code>
         is a popular way to run Node.js. It allows you to easily switch the Node.js version, and install new versions to try and easily rollback if something breaks, for example.
      </p>
      <p>
        It is also very useful to test your code with old Node.js versions.
      </p>
      <p>
        See 
        <a>
          https://github.com/nvm-sh/nvm
        </a>
         for more information about this option.
      </p>
      <p>
        In any case, when Node.js is installed you'll have access to the 
        <code
          class="code"
        >
          node
        </code>
         executable program in the command line.
      </p>
    </div>
    <div
      class="edit"
    >
      <a
        href="https://github.com/nodejs/nodejs.dev/edit/main/content/learn/test-path"
      >
        <span>
          Edit this page on GitHub
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pencil "
          data-icon="pencil"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
    <ul
      class="pagination"
    >
      <li>
        <a
          href="test-slug"
          rel="prev"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-left "
            data-icon="angles-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160zm352-160l-160 160c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L301.3 256 438.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0z"
              fill="currentColor"
            />
          </svg>
          Prev
        </a>
      </li>
      <li>
        <a
          href="test-slug"
          rel="next"
        >
          Next
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-right "
            data-icon="angles-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-left: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M470.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 256 265.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160zm-352 160l160-160c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L210.7 256 73.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0z"
              fill="currentColor"
            />
          </svg>
        </a>
      </li>
    </ul>
  </article>
</div>
`;

exports[`Article component should accept and render child components 1`] = `
<div>
  <article
    class="article"
  >
    <div
      class="header"
    >
      <h1
        class="headline"
      >
        test-title
      </h1>
      <p />
    </div>
    <div
      class="authorList"
    >
      Article Authors
      <ul>
        <li>
          <a
            aria-label="test-user1 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user1"
            rel="noopener noreferrer"
            style="margin-left: 0px;"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user1.png?size=60"
            />
          </a>
        </li>
        <li>
          <a
            aria-label="test-user2 Github - opens in new tab"
            class="link"
            href="https://github.com/test-user2"
            rel="noopener noreferrer"
            target="_blank"
          >
            <img
              alt=""
              src="https://github.com/test-user2.png?size=60"
            />
          </a>
        </li>
      </ul>
    </div>
    <details
      class="tableOfContents"
    >
      <summary>
        <strong>
          Table of Contents
        </strong>
      </summary>
      <ul>
        <li>
          <a
            href="#mock-title"
          >
            mock-title
          </a>
        </li>
        <li>
          <a
            href="#mock-title-2"
          >
            mock-title-2
          </a>
        </li>
      </ul>
    </details>
    <div>
      <p>
        Node.js can be installed in different ways. This post highlights the most common and convenient ones.
      </p>
      <p>
        Official packages for all the major platforms are available at 
        <a>
          https://nodejs.dev/download/
        </a>
        .
      </p>
      <p>
        One very convenient way to install Node.js is through a package manager. In this case, every operating system has its own.
      </p>
      <p>
        Other package managers for MacOS, Linux, and Windows are listed in 
        <a>
          https://nodejs.dev/download/package-manager/
        </a>
      </p>
      <p>
        <code
          class="code"
        >
          nvm
        </code>
         is a popular way to run Node.js. It allows you to easily switch the Node.js version, and install new versions to try and easily rollback if something breaks, for example.
      </p>
      <p>
        It is also very useful to test your code with old Node.js versions.
      </p>
      <p>
        See 
        <a>
          https://github.com/nvm-sh/nvm
        </a>
         for more information about this option.
      </p>
      <p>
        In any case, when Node.js is installed you'll have access to the 
        <code
          class="code"
        >
          node
        </code>
         executable program in the command line.
      </p>
    </div>
    <div>
      <div>
        test child JSX
      </div>
    </div>
    <div
      class="edit"
    >
      <a
        href="https://github.com/nodejs/nodejs.dev/edit/main/content/learn/test-path"
      >
        <span>
          Edit this page on GitHub
        </span>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pencil "
          data-icon="pencil"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1v32c0 8.8 7.2 16 16 16h32zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"
            fill="currentColor"
          />
        </svg>
      </a>
    </div>
    <ul
      class="pagination"
    >
      <li>
        <a
          href="test-slug"
          rel="prev"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-left "
            data-icon="angles-left"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-right: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160zm352-160l-160 160c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L301.3 256 438.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0z"
              fill="currentColor"
            />
          </svg>
          Prev
        </a>
      </li>
      <li>
        <a
          href="test-slug"
          rel="next"
        >
          Next
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-angles-right "
            data-icon="angles-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            style="margin-left: 5px;"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M470.6 278.6c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L402.7 256 265.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0l160-160zm-352 160l160-160c12.5-12.5 12.5-32.8 0-45.3l-160-160c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3L210.7 256 73.4 393.4c-12.5 12.5-12.5 32.8 0 45.3s32.8 12.5 45.3 0z"
              fill="currentColor"
            />
          </svg>
        </a>
      </li>
    </ul>
  </article>
</div>
`;
